### PR TITLE
mon/MonClient: weight-based mon selection

### PR DESCRIPTION
**Pull request by xiexingguo**
Monitors now support both priority and weight configurations.
Clients and daemons which need to connect to Monitors will now do Monitor selection
 in the following order:
  - first choose Monitor by priority, the lower value the more preferred
  - if there are still multiple targets, then choosing by their corresponding weight.
    If no valid(non-zero) weights is detected, all targets will be selected at random.

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
> see http://tracker.ceph.com/issues/5249 for the use case of priority.

@tchaikov  Thanks!
I add a comment to point that out in case any following contributors may try to make the same attempt.

Is that acceptable? If not, feel free to close :-)

File: src/mon/MonClient.cc [hunk 1]
```diff
@@ -687,21 +687,22 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
...
 {
-  uint16_t min_priority = std::numeric_limits<uint16_t>::max();
+  map<uint16_t, vector<unsigned>> rank_by_priority;
   for (const auto& m : monmap.mon_info) {
-    if (m.second.priority < min_priority) {
-      min_priority = m.second.priority;
-    }
+    rank_by_priority[m.second.priority].push_back(monmap.get_rank(m.first));
   }
   vector<unsigned> ranks;
```
Comment (liewegas) [inline src/mon/MonClient.cc]: Maybe what we really want is to make our selection of a rank proportional to it's priority.. i.e., a weight.  We would add ``weight`` instances of the rank to the vector, shuffle, and choose the first (unique) N of them.  Or maybe std:: has something to do weighted shuffle/selection).

Comment (xiexingguo) [reply src/mon/MonClient.cc]: Yup, that was __exact__ my initial thought.

Comment (tchaikov) [reply src/mon/MonClient.cc]: first, i don't understand why we want to rewrite the existing shuffling logic in a different way. and the weight and priority are different things in SRV record. i'd suggest avoid taking one as another.

**Comment by xiexingguo** [comment]
I haven't got a chance to use mon priorities before, so honestly I don't have a strong opinion on this topic.
But consider the following scenario:
```
mon.a       priority 0
mon.b       priority 1
mon.c       priority 2
```
following the current code logic seems all clients will be starving at hunting mon.a. I suppose it is the expected behavior, right?

Well, what I really want is:
```
mon.a       weight 2
mon.b       weight 1
mon.c       weight 1
```
e.g., we are assigning `mon.a` a higher weight because it has more powerful CPU, RAM, etc., and hence is 2x capable of handling incoming client connections. `mon.a`, `mon.b` and` mon.c` would get a 50%, 25% and 25% probabilities to handle clients OPs respectively as a result.

In other words I am expecting to handle the mon-priorities (or weights, whatever) much similar to the way we already did in CRUSH, which is something we are currently lack of?

**Comment by tchaikov** [comment]
> I suppose it is the expected behavior, right?

right.

> what I really want is:

probably you could file a tracker or implement it you are after this feature? but the change in this PR does not change the behavior as such or address your requirement. and the commit message failed to help me understand why we need the rewrite. the comment does help the posterity. but, personally, i presume developer should read the commit message of the code they are touching, though.

currently, only the priority of SRV is partially implemented for addressing the requirement of http://tracker.ceph.com/issues/5249 .

> In other words I am expecting to handle the mon-priorities (or weights, whatever) much similar to the way we already did in CRUSH, which is something we are currently lack of?

to model the monitor selection after CRUSH, or even just mimic it, would be a more complicated change, i guess. what i can think of is to (ab)use SRV record for this. and let the administrator to adjust the SRV settings for optimal distribution of monitor load. i admit we could offer more tooling in this area.

i think @jecluis has also been thinking about this -- to leverage the knowledge of topology for monitor selection.

**Comment by xiexingguo** [comment]
> but the change in this PR does not change the behavior as such or address your requirement. and the commit message failed to help me understand why we need the rewrite. the comment does help the posterity. but, personally, i presume developer should read the commit message of the code they are touching, though.

Aah, the commit message serves my original proposal `mon/MonClient: ignore mon priorities for now`.
I thought you'd be aware of that. Sorry that you aren't :joy: 

Deleting now!

**Comment by tchaikov** [comment]
> Sorry that you aren't

@xiexingguo i was looking at the commit message of your change, not the first comment of this PR. the reason why i am confused is that you rewrote the monitor selection without obvious reason IMHO. and your commit message puts

> mon/MonClient: add comment for the use case of mon-priority

apparently, you are doing more than this.

**Comment by tchaikov** [comment]
> In other words I am expecting to handle the mon-priorities (or weights, whatever) much similar to the way we already did in CRUSH, which is something we are currently lack of? 

see also the discussion in #23922

**Comment by xiexingguo** [comment]
@tchaikov Sorry for the late response.  I re-posted a weight-based mon selection version, just as @liewegas suggested:
>Maybe what we really want is to make our selection of a rank proportional to it's priority.. i.e., a weight. We would add weight instances of the rank to the vector, shuffle, and choose the first (unique) N of them. Or maybe std:: has something to do weighted shuffle/selection).

**Comment by xiexingguo** [comment]
retest this please

File: src/mon/MonmapMonitor.cc [hunk 2]
```diff
@@ -782,6 +782,35 @@ bool MonmapMonitor::prepare_command(MonOpRequestRef op)
...
+          !cmd_getval(g_ceph_context, cmdmap, "weight", value))) {
+      err = -EINVAL;
+      goto reply;
+    }
+    if (!pending_map.contains(name)) {
+      ss << "mon." << name << " does not exist";
+      err = -ENOENT;
+      goto reply;
+    }
+    if (value > max_value) {
```
Comment (tchaikov) [inline src/mon/MonmapMonitor.cc]: shall we enforce the minimum value also?

File: src/common/weighted_shuffle.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef CEPH_WEIGHTED_SHUFFLE_H
+#define CEPH_WEIGHTED_SHUFFLE_H
+
+#include <vector>
+#include <random>
+#include <algorithm>
+
+template <class T, class R>
+auto weighted_shuffle(std::vector<T> data, std::vector<int> weights, R &&gen)
```
Comment (tchaikov) [inline src/common/weighted_shuffle.h]: i guess this implementation comes from https://stackoverflow.com/questions/50221136/c-weighted-stdshuffle . so it's licensed under CC-BY-SA unless the author claims otherwise.

but it removes the nice things from the original version:

1. we can do the shuffle in-place, and no need to pass the shuffled vector by value
2. we should not pass the weight by value.

File: src/common/weighted_shuffle.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * This is free software; you can redistribute it and/or
```
Comment (tchaikov) [inline src/common/weighted_shuffle.h]: no need to repeat the license. see https://github.com/ceph/ceph/blob/master/CodingStyle#L73

File: src/common/weighted_shuffle.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef CEPH_WEIGHTED_SHUFFLE_H
+#define CEPH_WEIGHTED_SHUFFLE_H
+
+#include <vector>
+#include <random>
+#include <algorithm>
```
Comment (tchaikov) [inline src/common/weighted_shuffle.h]: nit, reorder the `#include`s alphabetically . see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/test/test_weighted_shuffle.cc [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    {'e', {0, 0, 0, 0, 0}}
+  }; // count each element appearing in each position
+  const int samples = 10000;
+  for (auto i = 0; i < samples; i++) {
+    auto shuffled = weighted_shuffle(data, weights, gen);
+    for (auto j = 0; j < data.size(); ++j)
+      ++frequency[shuffled[j]][j];
+  }
+
+  std::cout << "frequencies of each position for every element:" << std::endl;
```
Comment (tchaikov) [inline src/test/test_weighted_shuffle.cc]: i'd suggest avoid any output in unit test unless the test fails.

File: src/test/test_weighted_shuffle.cc [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  // verify each element gets a chance to stay at the header of array
+  // e.g., because we don't want to starve anybody!
+  for (auto& i : frequency) {
+    auto& counter = i.second;
+    EXPECT_TRUE(counter.front() > 0);
+  }
+
+  // verify the probability (of staying at the array header)
+  // are produced according to their corresponding weight
+  for (auto it = frequency.begin(); it != frequency.end(); it++) {
```
Comment (tchaikov) [inline src/test/test_weighted_shuffle.cc]: could use `std::is_sorted()`.

File: src/test/test_weighted_shuffle.cc [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    for (auto j : i.second)
+      std::cout << std::setw(6) << j;
+    std::cout << '\n';
+  }
+
+  // verify each element gets a chance to stay at the header of array
+  // e.g., because we don't want to starve anybody!
+  for (auto& i : frequency) {
+    auto& counter = i.second;
+    EXPECT_TRUE(counter.front() > 0);
```
Comment (tchaikov) [inline src/test/test_weighted_shuffle.cc]: `EXPECT_GE()` or use `EXPECT_TRUE(std:: all_of(...))`.

File: src/test/test_weighted_shuffle.cc [hunk 9]
```diff
@@ -0,0 +1,63 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * This is free software; you can redistribute it and/or
```
Comment (tchaikov) [inline src/test/test_weighted_shuffle.cc]: also, the major part of this file comes from https://wandbox.org/permlink/SuTYLHHG8brNNieq . probably we cannot claim it LGPL unless the author claims otherwise.

File: src/mon/MonmapMonitor.cc (hunk 2 above)
Comment (xiexingguo) [reply src/mon/MonmapMonitor.cc]: It should have been done by the command para type sanity checking:

`"name=priority,type=CephInt,range=0"`

File: src/mon/MonCommands.h [hunk 10]
```diff
@@ -463,6 +463,16 @@ COMMAND("mon set-addrs " \
 	"name=addrs,type=CephString",
 	"set the addrs (IPs and ports) a specific monitor binds to",
 	"mon", "rw")
+COMMAND("mon set-priority " \
```
[the remark this answers is not in the document]
Comment (xiexingguo) [reply src/mon/MonCommands.h]: > if you are fine, i can implement the proper priority support for SRV record in a separated PR as suggested above. and then, i think we are ready for the "mon set-priority" command.

Definitely :-)

**Comment by xiexingguo** [comment]
@tchaikov  Comment addressed expect the minimum value sanity checking.

File: src/mon/MonMap.h [hunk 11]
```diff
@@ -135,9 +137,10 @@ class MonMap {
   uint8_t min_mon_release = 0;
 
   void _add_ambiguous_addr(const string& name,
-			   entity_addr_t addr,
-			   int priority,
-			   bool for_mkfs=false);
+                           entity_addr_t addr,
+                           int priority = 0,
+                           int weight = 0,
```
Comment (tchaikov) [inline src/mon/MonMap.h]: i'd suggest avoid using the default parameters as much as possible here. as it's a private method. it'd be better if we can be more explicit.

File: src/common/weighted_shuffle.h [hunk 12]
```diff
@@ -0,0 +1,29 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+// https://wandbox.org/permlink/SuTYLHHG8brNNieq
```
Comment (tchaikov) [inline src/common/weighted_shuffle.h]: much like https://godbolt.org, wandbox is more of a service of compilation. the link you are referencing here cannot be tracked by to the author. it reads

> created by
> anonymous
> 10 months ago

we cannot take the risk of using code from nowhere by nobody. but yeah, if the author transferred the ownership of copyright to you, so you can have the Signed-off-by, that's another story.

so i think the safe approach is to reference the stackoverflow link : https://stackoverflow.com/questions/50221136/c-weighted-stdshuffle/50223540 . as 

1. it is explicitly licensed under CC-AS. see https://stackoverflow.com/help/licensing
1. we can at least track to its original author "Bob__". but sadly, i have no idea if or how we can have his Signed-off-by.

File: src/mon/MonClient.cc [hunk 13]
```diff
@@ -687,23 +688,49 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
...
+      weighted_shuffle(ranks, weights, gen);
     }
   }
-  std::random_device rd;
-  std::mt19937 rng(rd());
-  std::shuffle(ranks.begin(), ranks.end(), rng);
+  ldout(cct, 10) << __func__ << " ranks=" << ranks << dendl;
   unsigned n = cct->_conf->mon_client_hunt_parallel;
-  if (n == 0 || n > ranks.size()) {
+  if (n == 0 || n > ranks.size()) { // always choose first N
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: this comment is confusing. we are not choosing the "first N" here. we are connecting to *all* of the monitor candidates.

File: src/mon/MonClient.cc [hunk 14]
```diff
@@ -687,23 +688,49 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
...
+  if (!rank_by_priority.empty()) {
+    // pick mon by priority, lowest first
+    // see http://tracker.ceph.com/issues/5249 for the use case of priority
+    candidates = rank_by_priority.begin()->second;
   }
   vector<unsigned> ranks;
-  for (const auto& m : monmap.mon_info) {
-    if (m.second.priority == min_priority) {
-      ranks.push_back(monmap.get_rank(m.first));
+  if (candidates.size() > 1) {
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: what if there is only a single element in `candidates` ?

Comment (xiexingguo) [reply src/mon/MonClient.cc]: Good catch. All fixed.

**Comment by xiexingguo** [comment]
retest this please

File: src/mon/MonmapMonitor.cc (hunk 2 above)
Comment (tchaikov) [reply src/mon/MonmapMonitor.cc]: ic. can you make it `range=0|65535` ? see https://tools.ietf.org/html/rfc2782#page-3

Comment (xiexingguo) [reply src/mon/MonmapMonitor.cc]: Sure, added.

File: src/mon/MonMap.h [hunk 15]
```diff
@@ -56,15 +56,17 @@ struct mon_info_t {
    * the priority of the mon, the lower value the more preferred
    */
   uint16_t priority{0};
+  uint16_t weight{0};
```
Comment (tchaikov) [inline src/mon/MonMap.h]: i'd suggest use a non-zero weight. so the default behavior is to uniform distribution. and that's expected. also, no need to make this a special case.

File: src/test/test_weighted_shuffle.cc [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+  for (auto i = 0; i < samples; i++) {
+    weighted_shuffle(data, weights, gen);
+    for (size_t j = 0; j < data.size(); ++j)
+      ++frequency[data[j]][j];
+  }
+
+  // verify each element gets a chance to stay at the header of array
+  // e.g., because we don't want to starve anybody!
+  EXPECT_TRUE(std::all_of(frequency.begin(), frequency.end(),
+    [](std::pair<char, std::vector<int>> const& it) {
```
Comment (tchaikov) [inline src/test/test_weighted_shuffle.cc]: nit, could use `auto&` here.

File: src/mon/MonMap.h [hunk 17]
```diff
@@ -205,8 +208,8 @@ class MonMap {
    * @param addr Monitor's public address
    */
   void add(const string &name, const entity_addrvec_t &addrv,
-	   int priority=0) {
-    add(mon_info_t(name, addrv, priority));
+	   int priority=0, int weight=0) {
```
Comment (tchaikov) [inline src/mon/MonMap.h]: ditto. could define and use a default weight here. for instance, `10`.

**Comment by xiexingguo** [comment]
jenkins render docs

File: src/mon/MonMap.h [hunk 18]
```diff
@@ -56,15 +56,17 @@ struct mon_info_t {
    * the priority of the mon, the lower value the more preferred
    */
   uint16_t priority{0};
+  uint16_t weight{10};
```
Comment (tchaikov) [inline src/mon/MonMap.h]: magic number here. could you define a `constexpr` for the `DEFAULT_WEIGHT`?  and use it instead of "10".

File: src/mon/MonClient.cc [hunk 19]
```diff
@@ -687,21 +688,27 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
...
-  for (const auto& m : monmap.mon_info) {
-    if (m.second.priority == min_priority) {
-      ranks.push_back(monmap.get_rank(m.first));
+  if (!rank_by_priority.empty()) {
+    // pick mon by priority, lowest first
+    // see http://tracker.ceph.com/issues/5249 for the use case of priority
+    ranks = rank_by_priority.begin()->second;
+  }
+  if (ranks.size() > 1) {
+    vector<int> weights;
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: nit `s/int/uint16_t/` to match the type of weight.

File: src/test/CMakeLists.txt [hunk 20]
```diff
@@ -893,4 +893,9 @@ add_executable(unittest_any test_any.cc)
 add_ceph_unittest(unittest_any)
 target_link_libraries(unittest_any)
 
+# unittest_weighted_shuffle
+add_executable(unittest_weighted_shuffle test_weighted_shuffle.cc)
+add_ceph_unittest(unittest_weighted_shuffle)
+target_link_libraries(unittest_weighted_shuffle)
```
Comment (tchaikov) [inline src/test/CMakeLists.txt]: if it does not link against any additional libraries, why bother adding this line?

File: src/mon/MonClient.cc [hunk 21]
```diff
@@ -687,21 +688,27 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
...
-    if (m.second.priority < min_priority) {
-      min_priority = m.second.priority;
-    }
+    rank_by_priority[m.second.priority].push_back(monmap.get_rank(m.first));
   }
   vector<unsigned> ranks;
-  for (const auto& m : monmap.mon_info) {
-    if (m.second.priority == min_priority) {
-      ranks.push_back(monmap.get_rank(m.first));
+  if (!rank_by_priority.empty()) {
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: we will in a bigger trouble if `rank_by_priority.empty()`. could you instead
```c++
assert(!rank_by_priority.empty());
```
?

File: doc/rados/configuration/mon-lookup-dns.rst [hunk 22]
```diff
@@ -42,10 +42,15 @@ With those records now existing we can create the SRV TCP records with the name
...
+    _ceph-mon._tcp.example.com. 60 IN SRV 10 20 6789 mon1.example.com.
+    _ceph-mon._tcp.example.com. 60 IN SRV 10 30 6789 mon2.example.com.
+    _ceph-mon._tcp.example.com. 60 IN SRV 20 50 6789 mon3.example.com.
 
-In this case the Monitors are running on port *6789*, and their priority and weight are all *10* and *60* respectively.
+Now all Monitors are running on port *6789*, with priorities 10, 10, 20 and weights 20, 30, 50 respectively.
 
-The current implementation in clients and daemons will *only* respect the priority set in SRV records, and they will only connect to the monitors with lowest-numbered priority. The targets with the same priority will be selected at random.
+If specified, priorities and weights will be treated in the following order:
+1. priorities always take precedence over weights, the lower value the more preferred
```
Comment (tchaikov) [inline doc/rados/configuration/mon-lookup-dns.rst]: https://github.com/ceph/ceph/pull/26940#discussion_r266862680

use 
```rst
#. first
#. second
``` 
for ordered items

File: doc/rados/configuration/mon-lookup-dns.rst [hunk 23]
```diff
@@ -42,10 +42,15 @@ With those records now existing we can create the SRV TCP records with the name
...
-    _ceph-mon._tcp.example.com. 60 IN SRV 10 60 6789 mon3.example.com.
+    _ceph-mon._tcp.example.com. 60 IN SRV 10 20 6789 mon1.example.com.
+    _ceph-mon._tcp.example.com. 60 IN SRV 10 30 6789 mon2.example.com.
+    _ceph-mon._tcp.example.com. 60 IN SRV 20 50 6789 mon3.example.com.
 
-In this case the Monitors are running on port *6789*, and their priority and weight are all *10* and *60* respectively.
+Now all Monitors are running on port *6789*, with priorities 10, 10, 20 and weights 20, 30, 50 respectively.
 
-The current implementation in clients and daemons will *only* respect the priority set in SRV records, and they will only connect to the monitors with lowest-numbered priority. The targets with the same priority will be selected at random.
+If specified, priorities and weights will be treated in the following order:
```
Comment (tchaikov) [inline doc/rados/configuration/mon-lookup-dns.rst]: add a newline here.

Comment (tchaikov) [inline doc/rados/configuration/mon-lookup-dns.rst]: please remove "If specified".

monitor clients choose monitor by referencing the SRV records. If a cluster has multiple monitor SRV records with the same priority value, clients and daemons will load balance the connections to monitors in proportion to the values of the SRV weight fields.

File: src/mon/MonClient.cc (hunk 21 above)
Comment (xiexingguo) [reply src/mon/MonClient.cc]: seems the original code logic should respect an empty monmap. I am not sure this will cause any other potential issue. Updated as you wish, though.

**Comment by tchaikov** [comment]
see
- http://pulpito.ceph.com/kchai-2019-03-21_02:00:24-rados-wip-kefu-testing-2019-03-21-0010-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-03-22_03:22:17-rados-wip-kefu-testing-2019-03-21-0010-distro-basic-mira/

some of the failures

- the upgrade failures should be addressed by #27112
- the test_python.sh failures should be addressed by #27111

**Comment by xiexingguo** [comment]
@tchaikov Thanks!
